3_time_series/Statistical_analysis.md

```diff
@@ -0,0 +1,749 @@
+def predict_previous_value(dataframe, split):
+    return dataframe[split + relativedelta(months=-1): ][:-1].shift(1, freq='M')
+```
+
+```python
+def exponentialSmoothing(dataframe, split, verbose=False):
+    model_fit = []
+    predictions =  []
+    for i, cur_date in enumerate(pd.date_range(split,  dataframe.index[-1], freq='MS')):
+        if(verbose): print(i, " iteration")
+        model = ExponentialSmoothing(dataframe[:cur_date - timedelta(days=1)], seasonal_periods = 12, trend='add', seasonal='add')
+        fit = model.fit()
+        pred = fit.predict(start=cur_date, end=cur_date) 
+        
+        model_fit.append(fit)
+        predictions.append(pred)
+    
+    predictions_df = pd.DataFrame(pd.concat(predictions, axis=0), columns=['Num_Passengers'])
+    return model_fit, predictions_df
+```
+
+```python
+def linear_regression_with_season(x, y):
+    if len(y)<12:
+        raise ValueError("This prediction require more then year of test data")
+    model = LinearRegression()
+    fit = model.fit(x, y)
+    lin_pred = fit.predict([x[-1,:]+1])
+    pred = lin_pred[0][0] + (y[-12][0] - fit.predict([x[-12,:]])[0][0])
+    return fit, pred
+```
+
+```python
+def fit_linear_regression_with_season(dataframe, split, verbose=False):
+    model_fit = []
+    predictions =  []
+    prediction_range = pd.date_range(split,  dataframe.index[-1], freq='MS')
+    for i, cur_date in enumerate(prediction_range):
+        if(verbose): print(i, " iteration")
+        data = dataframe[:cur_date - timedelta(days=1)]['Num_Passengers'].values.reshape(-1, 1)
+        index = np.arange(len(data)).reshape(-1, 1)
+        fit, pred = linear_regression_with_season(index, data)
+        
+        model_fit.append(fit)
+        predictions.append(pred)
+    predictions_df = pd.DataFrame({'Month': prediction_range,
+                   'Num_Passengers': predictions})
+    predictions_df = predictions_df.set_index('Month')
+    
+    return model_fit, predictions_df
+```
+
+```python
+def linear_regression(x, y, predict):
+    model = LinearRegression()
+    fit = model.fit(x, y)
+    pred = fit.predict(predict)[0]
+    return fit, pred
+```
+
+```python
+def fit_linear_regression_with_12_lag(dataframe, split, verbose=False):
+    model_fit = []
+    predictions =  []
+    prediction_range = pd.date_range(split,  dataframe.index[-1], freq='MS')
+    for i, cur_date in enumerate(prediction_range):
+        if(verbose): print(i, " iteration")
+        train_data = dataframe[:cur_date].reset_index(drop=True)
+        train_data["lag_12"] = train_data["Num_Passengers"].shift(12)
+        train_data = train_data[12:]
+        train_data["lag_1"] = train_data["Num_Passengers"].shift(1)
+        train_data = train_data[1:]
+        
+        train_col = ["lag_1", "lag_12"]
+        fit, pred = linear_regression(train_data[train_col].iloc[:-1, :], train_data["Num_Passengers"].iloc[:-1], 
+                                     train_data[train_col].iloc[-1:, :])
```

Comment:
there is a bug with columns in here, right? for training you take `.iloc[:-1, :]` and while predicting `.iloc[-1:, :]`